apis/namespaced/common/common_methods.go

```diff
@@ -0,0 +1,55 @@
+package common
+
+import (
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// HasApplicationRef returns true if the InstanceData has an ApplicationRef set.
+// This indicates that the instance is linked to a separate Application resource.
+func (spec *InstanceData) HasApplicationRef() bool {
+	return spec.ApplicationRef != nil
+}
+
+// HasApplicationData returns true if the InstanceData contains inline ApplicationData.
+// This indicates that the instance stores application information directly.
+func (spec *InstanceData) HasApplicationData() bool {
+	return spec.ApplicationData != nil
+}
+
+// GetObjectKeyFromApplicationRef returns the Kubernetes ObjectKey corresponding
+// to the ApplicationRef in the InstanceData. This can be used to fetch the referenced
+// Application from the cluster.
+func (spec *InstanceData) GetObjectKeyFromApplicationRef() client.ObjectKey {
+	return spec.ApplicationRef.ToObjectKey()
```

Comment:
it's better to returns a zero ObjectKey if spec or ApplicationRef is nil
```suggestion
    if spec == nil || spec.ApplicationRef == nil {
        return client.ObjectKey{}
    }
	return spec.ApplicationRef.ToObjectKey()
```